exporters/otlp/otlpmetric/otlpmetrichttp/client.go

```diff
@@ -0,0 +1,286 @@
+	DialContext: (&net.Dialer{
+		Timeout:   30 * time.Second,
+		KeepAlive: 30 * time.Second,
+	}).DialContext,
+	ForceAttemptHTTP2:     true,
+	MaxIdleConns:          100,
+	IdleConnTimeout:       90 * time.Second,
+	TLSHandshakeTimeout:   10 * time.Second,
+	ExpectContinueTimeout: 1 * time.Second,
+}
+
+// newClient creates a new HTTP metric client.
+func newClient(opts ...Option) (otlpmetric.Client, error) {
+	cfg := oconf.NewHTTPConfig(asHTTPOptions(opts)...)
+
+	httpClient := &http.Client{
+		Transport: ourTransport,
+		Timeout:   cfg.Metrics.Timeout,
+	}
+	if cfg.Metrics.TLSCfg != nil {
+		transport := ourTransport.Clone()
+		transport.TLSClientConfig = cfg.Metrics.TLSCfg
+		httpClient.Transport = transport
+	}
+
+	format := "https://%s%s"
+	if cfg.Metrics.Insecure {
+		format = "http://%s%s"
+	}
+	rawURL := fmt.Sprintf(format, cfg.Metrics.Endpoint, cfg.Metrics.URLPath)
+	// Body is set when this is cloned during upload.
+	req, err := http.NewRequest(http.MethodPost, rawURL, http.NoBody)
+	if err != nil {
+		return nil, err
+	}
+
+	if n := len(cfg.Metrics.Headers); n > 0 {
+		for k, v := range cfg.Metrics.Headers {
+			req.Header.Set(k, v)
+		}
+	}
+	req.Header.Set("Content-Type", "application/x-protobuf")
+
+	return &client{
+		compression: Compression(cfg.Metrics.Compression),
+		req:         req,
+		requestFunc: cfg.RetryConfig.RequestFunc(evaluate),
+		httpClient:  httpClient,
+	}, nil
+}
+
+// ForceFlush does nothing, the client holds no state.
+func (c *client) ForceFlush(ctx context.Context) error { return ctx.Err() }
+
+// Shutdown shuts down the client, freeing all resources.
+func (c *client) Shutdown(ctx context.Context) error {
+	// The otlpmetric.Exporter synchronizes access to client methods and
+	// ensures this is called only once. The only thing that needs to be done
+	// here is to release any computational resources the client holds.
+
+	c.requestFunc = nil
+	c.httpClient = nil
+	return ctx.Err()
+}
+
+// UploadMetrics sends protoMetrics to the connected endpoint.
+//
+// Retryable errors from the server will be handled according to any
+// RetryConfig the client was created with.
+func (c *client) UploadMetrics(ctx context.Context, protoMetrics *metricpb.ResourceMetrics) error {
+	// The otlpmetric.Exporter synchronizes access to client methods, and
+	// ensures this is not called after the Exporter is shutdown. Only thing
+	// to do here is send data.
+
+	pbRequest := &colmetricpb.ExportMetricsServiceRequest{
+		ResourceMetrics: []*metricpb.ResourceMetrics{protoMetrics},
+	}
+	body, err := proto.Marshal(pbRequest)
+	if err != nil {
+		return err
+	}
+	request, err := c.newRequest(ctx, body)
+	if err != nil {
+		return err
+	}
+
+	return c.requestFunc(ctx, func(iCtx context.Context) error {
+		select {
+		case <-iCtx.Done():
+			return iCtx.Err()
+		default:
+		}
+
+		request.reset(iCtx)
+		resp, err := c.httpClient.Do(request.Request)
+		if err != nil {
+			return err
+		}
+
+		var rErr error
```

Comment:
There's discussion in #3106, in which I would recommend restructuring the way errors are handled in this block of code. I would be glad to follow-up with the same feedback here _after this PR_.

Reply:
> There's discussion in #3106, in which I would recommend restructuring the way errors are handled in this block of code. I would be glad to follow-up with the same feedback here _after this PR_.

Absolutely, that sounds like a good plan to me :+1: